test/azul_test_case.py

```diff
@@ -20,3 +60,86 @@ def setUpClass(cls) -> None:
     def tearDownClass(cls) -> None:
         boto3.session.Session.get_credentials = cls.get_credentials_orig
         super().tearDownClass()
+
+
+class Test:
+    # Keep this test case out of the main namespace
+
+    class LoggingTestCase(TestCase):
+        """A test case which logs its calls."""
+
+        def __init__(self, events):
+            super(Test.LoggingTestCase, self).__init__('test')
+            self.events = events
+
+        def setUp(self):
+            self.events.append('setUp')
+
+        def test(self):
+            self.events.append('test')
+
+        def tearDown(self):
+            self.events.append('tearDown')
+
+
+class TestAlwaysTearDownTestCase(TestCase):
+
+    def test_regular_execution_order(self):
+        for always_tear_down in (False, True):
+            events = []
+            tear_down_parent = [AlwaysTearDownTestCase] if always_tear_down else []
+            with self.subTest(always_tear_down=always_tear_down):
```

Comment:
I'd move this up to right after the `for` loop. That way it doesn't disrupt the flow.